src/Interactive/Features/Interactive/Core/InteractiveHost.RemoteService.cs

```diff
@@ -113,37 +115,38 @@ private void ReadOutput(bool error)
 
             internal void Dispose(bool joinThreads)
             {
-                // set _disposing so that we don't attempt restart the host anymore:
-                _disposing = true;
-
-                InitiateTermination(Process, _processId);
-
-                // only tests require joining the threads, so we can wait synchronously
-                if (joinThreads)
+                // There can be a call from host initiated from OnProcessExit. 
+                // This check on the beginning helps to avoid a reentrancy.
+                if (_processExitHandlerStatus == ProcessExitHandlerStatus.Hooked)
```

Comment:
Is the problem here that this can deadlock because if the _host.OnProcessExited() call is happening above and that's still holding the lock, and it tries to call back into Dispose(), that will deadlock? Why can't we call _host.OnProcessExited() outside the lock?

I guess a better comment here would be appreciated. The use of `volatile` here is generally scary. I think it works, but that doesn't make it less scary.

Reply:
(this question is best answered by just adding more comments so the code is clearer vs. answering my question in GitHub)

Reply:
Thank you, Jason! I added more comments to clarify the scenario. 

Maybe there can be an option to avoid using volatile and removing the _host.OnProcessExited() outside of the lock. We could not see these options so far. The problem we fixed that the host is being disposed during _host.OnProcessExited() execution. So, the fix is to protect it from the dispose.

---
In reply to: [163994639](https://github.com/dotnet/roslyn/pull/24058#discussion_r163994639) [](ancestors = 163994639)